automatic-config-updates/README.md

```diff
@@ -0,0 +1,71 @@
+With this in mind, I consider 2) to be a must-have, while 1) to be a nice-to-have.
+
+On the other hand, 1) has the advantage that it can fix the config of the user before it is merged to main if they try to use a deprecated feature, which 2) can't really handle (the config would be merged and then we would create a PR with the correction, so a correct config will end up on the main branch in the end).
+
+## Implementation approach
+
+Regardless of which of the following approaches ends up being used, the process of updating the config will always be the following:
+
+1. Get the config, determine if transformations need to be done
+2. Pull the repository
+3. Make transformations to the config
+4. Create a pull request (or add a commit to a PR) with the transformations
+
+Let's take a closer look at the more "problematic" parts of this pipeline.
+
+### Getting the config
+
+There seem to be three main approaches to getting the config:
+
+1. In case we do the modifications by adding commits to PRs, the config should come from the PR.
+2. Get the list of projects by querying `/api/projects` and then find configs in them.
+3. Add a new database table mapping project to its config content on main.
+
+Let's focus on comparing 2) and 3) since those are relevant to the periodic updates.
+Getting the list of projects with Packit enabled is straight-forward. However, we would have to do quite a lot of queries to Github to get all the configs (we could even run into Github's request quota).
+Moreover, with this approach, we can't store any additional information about the config, which could perhaps be useful.
+
+On the other hand, 3) would spread the requests to Github to get the current config across time. We would basically have to create an implicit action that triggers on merge to main/master and fetches the current config to the database. Hence, we wouldn't be making a lot of requests to Github at the same time, but rather over time.
+This approach also facilitates storing additional data about the configs, for example which transformations have already been applied (see the following section).
+Such additional information could allow us to detect that there are no transformations to do without even checking the config (resulting in faster overall updates).
+
+The main downside of storing the configuration in the database are the space requirements.
+However, this should hopefully not be a problem, since the config is usually not that long (but it is still something to keep in mind).
+To slightly improve the situation, the configs can be stored in a separate storage.
+We could store the raw content as a JSON in a document NoSQL database (e.g. MongoDB) and just map project to object ID in Mongo in postgres.
+This would also allow us to more easily query the existing configs for information about them (using an aggregation pipeline).
```

Comment:
"more easily" - by using two different databases instead of one?

PostgreSQL supports [native JSON data type](https://www.postgresql.org/docs/13/datatype-json.html) since version 9.2.
It provides many [functions and operators for manipulating JSON data](https://www.postgresql.org/docs/13/functions-json.html).
[SQLAlchemy seems to be able to utilize that as well](https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#table-valued-functions).

What is "aggregation pipeline"?

Reply:
that would be very effective if we could just dump those configs in psql

Reply:
Oh, cool, didn't know that postgres had such extensive support for json. Looks like with this, just storing it in postgres should be sufficient. I will update the doc